Fix notifications appearing during gameplay

Aergwyn opened the pull request:
+ Resolves #2649
+ fix `Toolbar` being toggleable when it shouldn't be able to, e.g. `Intro` (can't find the issue D:)
+ allow opening overlays in `MenuState.Initial` again (resolves #2643)

This isn't *complete* as it's missing `OverlayActivation.UserTriggered` but I'm not sure how to do that yet and I thought that this PR could either be used
a) for feedback how to accomplish that or
b) may be valid before adding the last bit of functionality.

Aergwyn committed:
add OverlayActivation enum

+ fix Toolbar being toggleable when it shouldn't be able to
+ allow opening overlays in MenuState.Initial again

peppy committed:
Merge branch 'master' into overlay-activation

peppy commented on osu.Game/Graphics/Containers/OsuFocusedOverlayContainer.cs:
```diff
@@ -16,13 +17,16 @@ public class OsuFocusedOverlayContainer : FocusedOverlayContainer
 
-        private readonly BindableBool allowOpeningOverlays = new BindableBool(true);
+        /// <summary>
+        /// Defaults to <see cref="OverlayActivation.All"/> so that the overlay works even if BDL couldn't pass an <see cref="OsuGame"/>.
```
I don't think the default needs to be explained.

peppy commented on osu.Game/Screens/Menu/ButtonSystem.cs:
```diff
@@ -332,19 +329,19 @@ private void updateLogoState(MenuState lastState = MenuState.Initial)
                 case MenuState.Exit:
                 case MenuState.Initial:
                     logoTracking = false;
+                    allowOverlays.Value = OverlayActivation.Disabled;
```
I'm pretty sure we don't want this `Disabled`. If anything, I'd uncomment `UserTriggered` and set that here so we aren't outright disabling for the time being (even if the user triggered support is not in yet).

peppy replied:
Also this is probably a use-case for hiding overlays *once* via a method call, but not actually disabling. Without the `All` set below in the delayed delegate (which is bound to cause weirdness).

Aergwyn replied:
Regarding the first comment: It completely falls flat as I noticed that any logic for that in `ButtonSystem` is now unnecessary.
Regarding the second comment: Every time the `MenuState` changes the visibility of the overlays is reevaluated as seen [here](https://github.com/Aergwyn/osu/blob/707af0209732ef29de79d4417c2aab9892ff8f89/osu.Game/Screens/Menu/MainMenu.cs#L67).

peppy commented on osu.Game/Screens/OsuScreen.cs:
```diff
@@ -103,8 +109,18 @@ private void load(OsuGameBase game, OsuGame osuGame, AudioManager audio)
                 Ruleset.BindTo(osuGame.Ruleset);
-                hideOverlaysOnEnter.BindTo(osuGame.HideOverlaysOnEnter);
-                allowOpeningOverlays.BindTo(osuGame.AllowOpeningOverlays);
+                AllowOverlays.BindTo(osuGame.AllowOverlays);
```
This is seriously dangerous. You're both binding to the global bindable and changing the value of it in BDL async methods. This means if a screen is preloaded without being immediately pushed, it is potentially changing the state of this variable.

Aergwyn committed:
apply feedback

-don't directly set AllowOverlay Bindable
this should be done specifically where needed
-remove AllowOverlay Bindable from ButtonSystem
-remove unnecessary xmldoc

Aergwyn committed:
Merge branch 'master' into overlay-activation

Aergwyn committed:
remove white space

I think

peppy committed:
Merge branch 'master' into overlay-activation

peppy committed:
Merge branch 'master' into overlay-activation

peppy commented on osu.Game/Screens/Menu/MainMenu.cs:
```diff
@@ -64,6 +63,8 @@ public MainMenu()
                 sideFlashes = new MenuSideFlashes(),
             };
+
+            buttons.StateChanged += state => UpdateOverlayStates?.Invoke();
```
this can't be

peppy commented on osu.Game/Screens/Menu/ButtonSystem.cs:
```diff
@@ -331,9 +321,6 @@ private void updateLogoState(MenuState lastState = MenuState.Initial)
                     {
                         logoTracking = false;
 
-                        hideOverlaysOnEnter.Value = true;
```
moving this out has caused a regression in the menu animation (timings have changed)

peppy commented on osu.Game/Screens/OsuScreen.cs:
```diff
@@ -103,8 +110,18 @@ private void load(OsuGameBase game, OsuGame osuGame, AudioManager audio)
+                    if (HideOverlaysOnEnter)
+                    {
+                        osuGame.CloseAllOverlays();
+                        osuGame.Toolbar.State = Visibility.Hidden;
```
why is the toolbar state handled here rather than in `OsuGame`?

peppy committed:
Use UserTriggered in Player

peppy committed:
Rename variables

peppy committed:
Ensure notifications don't appear during UserTriggered mode

Closes #2640.

peppy committed:
Rewrite much state logic

Aergwyn committed:
match Bindable names

peppy committed:
Fix missing null checks

peppy committed:
Merge branch 'master' into overlay-activation
